platform/pulp/plugin/changeset/iterator.py

```diff
@@ -0,0 +1,215 @@
+import itertools
+
+from logging import getLogger
+
+from django.db.models import Q
+
+from pulp.download import Download
+
+from .model import RemoteArtifact
+
+
+log = getLogger(__name__)
+
+
+class ContentIterator:
+    """
+    Iterate a generator of `RemoteContent` and replace associated content models
+    With models fetched from the DB (when found).
+    """
+
+    # Number of keys in each batched query.
+    BATCH = 1024
+
+    @staticmethod
+    def _batch_q(content):
+        """
+        Build a query for the specified batch of content.
+
+        Args:
+            content (tuple): A batch of content.  Each is: RemoteContent.
+
+        Returns:
+            Q: The built query.
+        """
+        query = Q()
+        for c in content:
+            query = query | Q(**c.key)
+        return query
+
+    def __init__(self, content):
+        """
+        Args:
+            content (generator): A generator of RemoteContent.
+        """
+        self.content = content
+
+    def _batch_content(self, size=BATCH):
+        """
+        Slice the content into batches.
+
+        Args:
+            size (int, optional): The number of content in each batch.
+
+        Yields:
+            tuple: Each batch of: RemoteContent.
+        """
+        content = (c for c in self.content)
+        while True:
+            batch = tuple(itertools.islice(content, 0, size))
+            if batch:
+                yield batch
+            else:
+                return
+
+    def _collated_content(self):
+        """
+        Collate each batch of content into lists by model.
+
+        Yields:
+            dict: A dictionary of {model_class: [content,]}
+                Each content is: RemoteContent.
+        """
+        for batch in self._batch_content():
+            collated = {}
+            for content in batch:
+                _list = collated.setdefault(type(content.model), list())
+                _list.append(content)
+            yield collated
+
+    def _fetch(self):
+        """
+        Fetch each batch of collated content.
+
+        Yields:
+            tuple: (content, fetched).
+                The content is a list of RemoteContent.
+                The fetched is a dictionary of fetched content models keyed by natural key.
+        """
+        for collated in self._collated_content():
+            for model, content in collated.items():
+                fields = (f.name for f in model.natural_key_fields)
```

Comment:
Using a generator comprehension here strikes me as a little funny. Not wrong, but if it we me I'd use a list comprehension since it'll probably get evaluated 4 lines below anyway in the call to ``q_set.only(``, and we don't gain anything here by delaying evaluation.